File: python/lance_namespace/tests/test_dir.py
"""
Tests for DirectoryNamespace implementation.
"""
import tempfile
import shutil
import os
from pathlib import Path

import pytest

from lance_namespace.dir import DirectoryNamespace
from lance_namespace_urllib3_client.models import (
    CreateNamespaceRequest,
    ListNamespacesRequest,
    DescribeNamespaceRequest,
    DropNamespaceRequest,
    NamespaceExistsRequest,
    ListTablesRequest,
    CreateTableRequest,
    CreateTableResponse,
    DropTableRequest,
    DropTableResponse,
    DescribeTableRequest,
    DescribeTableResponse,
    JsonArrowSchema,
    JsonArrowField,
    JsonArrowDataType,
)


class TestDirectoryNamespace:
    
    def setup_method(self):
        """Set up test environment."""
        self.temp_dir = tempfile.mkdtemp()
        self.namespace = DirectoryNamespace(root=self.temp_dir)
    
    def teardown_method(self):
        """Clean up test environment."""
        if os.path.exists(self.temp_dir):
            shutil.rmtree(self.temp_dir)
    
    def create_test_schema(self):
        """Create a test schema with id (int32) and name (string) fields."""
        # Create int32 type
        int_type = JsonArrowDataType(type="int32")
        
        # Create string type  
        string_type = JsonArrowDataType(type="utf8")
        
        # Create id field
        id_field = JsonArrowField(
            name="id",
            type=int_type,
            nullable=False
        )
        
        # Create name field
        name_field = JsonArrowField(
            name="name",
            type=string_type,
            nullable=True
        )
        
        # Create schema
        schema = JsonArrowSchema(fields=[id_field, name_field])
        return schema
    
    def test_init_with_absolute_path(self):
        """Test initialization with absolute path."""
        namespace = DirectoryNamespace(root=self.temp_dir)
        assert namespace is not None
    
    def test_init_with_file_uri(self):
        """Test initialization with file:// URI."""
        namespace = DirectoryNamespace(root=f"file://{self.temp_dir}")
        assert namespace is not None
    
    def test_init_with_relative_path(self):
        """Test initialization with relative path."""
        namespace = DirectoryNamespace(root="./test-namespace")
        assert namespace is not None
    
    def test_init_without_root(self):
        """Test initialization without root uses current directory."""
        # Should not raise error, uses current directory
        namespace = DirectoryNamespace()
        assert namespace is not None
    
    def test_create_table(self):
        """Test creating a table."""
        request = CreateTableRequest(
            id=["test_table"],
            schema=self.create_test_schema()
        )
        
        response = self.namespace.create_table(request, b"")
        assert response.location is not None
        assert "test_table" in response.location
        assert response.version == 1
        
        # Verify Lance dataset was created (check for _versions directory)
        table_dir = Path(self.temp_dir) / "test_table"
        assert table_dir.exists()
        assert table_dir.is_dir()
        
        versions_dir = table_dir / "_versions"
        assert versions_dir.exists()
        assert versions_dir.is_dir()
    
    def test_list_tables(self):
        """Test listing tables."""
        # Create some tables
        for table_name in ["table1", "table2", "table3"]:
            request = CreateTableRequest(
                id=[table_name],
                schema=self.create_test_schema()
            )
            self.namespace.create_table(request, b"")
        
        # List tables
        request = ListTablesRequest()
        response = self.namespace.list_tables(request)
        
        assert len(response.tables) == 3
        assert set(response.tables) == {"table1", "table2", "table3"}
    
    def test_drop_table(self):
        """Test dropping a table."""
        # First create a table
        create_request = CreateTableRequest(
            id=["test_table"],
            schema=self.create_test_schema()
        )
        self.namespace.create_table(create_request, b"")
        
        table_dir = Path(self.temp_dir) / "test_table"
        assert table_dir.exists()
        
        # Drop the table
        drop_request = DropTableRequest()
        drop_request.id = ["test_table"]
        response = self.namespace.drop_table(drop_request)
        
        assert response is not None
        # Verify table directory was removed
        assert not table_dir.exists()
    
    def test_empty_list_tables(self):
        """Test listing tables when none exist."""
        request = ListTablesRequest()
        response = self.namespace.list_tables(request)
        
        assert response.tables == []
    
    def test_namespace_operations_not_supported(self):
        """Test that namespace operations raise NotImplementedError."""
        
        # Test CreateNamespace
        with pytest.raises(NotImplementedError, match="flat list of tables"):
            self.namespace.create_namespace(CreateNamespaceRequest())
        
        # Test ListNamespaces
        with pytest.raises(NotImplementedError, match="flat list of tables"):
            self.namespace.list_namespaces(ListNamespacesRequest())
        
        # Test DescribeNamespace
        with pytest.raises(NotImplementedError, match="flat list of tables"):
            self.namespace.describe_namespace(DescribeNamespaceRequest())
        
        # Test DropNamespace
        with pytest.raises(NotImplementedError, match="flat list of tables"):
            self.namespace.drop_namespace(DropNamespaceRequest())
        
        # Test NamespaceExists
        with pytest.raises(NotImplementedError, match="flat list of tables"):
            self.namespace.namespace_exists(NamespaceExistsRequest())
    
    def test_describe_table(self):
        """Test describing a table."""
        # First create a table
        create_request = CreateTableRequest(
            id=["test_table"],
            schema=self.create_test_schema()
        )
        self.namespace.create_table(create_request, b"")
        
        # Now describe the table
        describe_request = DescribeTableRequest()
        describe_request.id = ["test_table"]
        response = self.namespace.describe_table(describe_request)
        
        assert response is not None
        assert response.location is not None
        assert "test_table" in response.location
        assert self.temp_dir in response.location
    
    def test_describe_nonexistent_table(self):
        """Test describing a nonexistent table."""
        describe_request = DescribeTableRequest()
        describe_request.id = ["nonexistent_table"]
        
        with pytest.raises(RuntimeError, match="Table does not exist"):
            self.namespace.describe_table(describe_request)
    
    def test_create_table_invalid_id(self):
        """Test creating table with invalid ID."""
        request = CreateTableRequest(
            id=[],  # Empty ID
            schema=self.create_test_schema()
        )
        
        with pytest.raises(ValueError, match="table ID must have exactly 1 level"):
            self.namespace.create_table(request, b"")

File: rust/lance-namespace/tests/test_dir.rs
//! Tests for DirectoryNamespace implementation

use std::collections::HashMap;

use lance_namespace::{DirectoryNamespace, LanceNamespace, NamespaceError};
use lance_namespace_reqwest_client::models::{
    CreateTableRequest, DescribeTableRequest, DropTableRequest, JsonArrowDataType,
    JsonArrowField, JsonArrowSchema, ListTablesRequest, TableExistsRequest,
};
use tempfile::TempDir;

/// Helper to create a test DirectoryNamespace with a temporary directory
async fn create_test_namespace() -> (DirectoryNamespace, TempDir) {
    let temp_dir = TempDir::new().unwrap();
    let mut properties = HashMap::new();
    properties.insert(
        "root".to_string(),
        temp_dir.path().to_string_lossy().to_string(),
    );

    let namespace = DirectoryNamespace::new(properties).unwrap();
    (namespace, temp_dir)
}

/// Helper to create a simple test schema
fn create_test_schema() -> JsonArrowSchema {
    let int_type = JsonArrowDataType::new("int32".to_string());
    let string_type = JsonArrowDataType::new("utf8".to_string());

    let id_field = JsonArrowField {
        name: "id".to_string(),
        r#type: Box::new(int_type.clone()),
        nullable: false,
        metadata: None,
    };

    let name_field = JsonArrowField {
        name: "name".to_string(),
        r#type: Box::new(string_type),
        nullable: true,
        metadata: None,
    };

    JsonArrowSchema {
        fields: vec![id_field, name_field],
        metadata: None,
    }
}

/// Test creating a table
#[tokio::test]
async fn test_create_table() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    let mut request = CreateTableRequest::new();
    request.id = Some(vec!["test_table".to_string()]);
    request.schema = Some(Box::new(create_test_schema()));

    let response = namespace
        .create_table(request, bytes::Bytes::new())
        .await
        .unwrap();

    assert!(response.location.is_some());
    assert!(response.location.unwrap().ends_with("test_table.lance"));
    assert_eq!(response.version, Some(1));
}

/// Test creating a table without schema should fail
#[tokio::test]
async fn test_create_table_without_schema() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    let mut request = CreateTableRequest::new();
    request.id = Some(vec!["test_table".to_string()]);
    // No schema provided

    let result = namespace.create_table(request, bytes::Bytes::new()).await;
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_string()
        .contains("Schema is required"));
}

/// Test creating a table with invalid ID should fail
#[tokio::test]
async fn test_create_table_with_invalid_id() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    // Test with empty ID
    let mut request = CreateTableRequest::new();
    request.id = Some(vec![]);
    request.schema = Some(Box::new(create_test_schema()));

    let result = namespace.create_table(request, bytes::Bytes::new()).await;
    assert!(result.is_err());

    // Test with multi-level ID
    let mut request = CreateTableRequest::new();
    request.id = Some(vec!["namespace".to_string(), "table".to_string()]);
    request.schema = Some(Box::new(create_test_schema()));

    let result = namespace.create_table(request, bytes::Bytes::new()).await;
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_string()
        .contains("single-level table IDs"));
}

/// Test creating a table with mismatched location
#[tokio::test]
async fn test_create_table_with_wrong_location() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    let mut request = CreateTableRequest::new();
    request.id = Some(vec!["test_table".to_string()]);
    request.schema = Some(Box::new(create_test_schema()));
    request.location = Some("/wrong/path/table.lance".to_string());

    let result = namespace.create_table(request, bytes::Bytes::new()).await;
    assert!(result.is_err());
    assert!(result.unwrap_err().to_string().contains("must be at location"));
}

/// Test listing tables
#[tokio::test]
async fn test_list_tables() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    // Initially, no tables
    let request = ListTablesRequest::new();
    let response = namespace.list_tables(request).await.unwrap();
    assert_eq!(response.tables.len(), 0);

    // Create a table
    let mut create_request = CreateTableRequest::new();
    create_request.id = Some(vec!["table1".to_string()]);
    create_request.schema = Some(Box::new(create_test_schema()));
    namespace
        .create_table(create_request, bytes::Bytes::new())
        .await
        .unwrap();

    // Create another table
    let mut create_request = CreateTableRequest::new();
    create_request.id = Some(vec!["table2".to_string()]);
    create_request.schema = Some(Box::new(create_test_schema()));
    namespace
        .create_table(create_request, bytes::Bytes::new())
        .await
        .unwrap();

    // List tables should return both
    let request = ListTablesRequest::new();
    let response = namespace.list_tables(request).await.unwrap();
    let tables = response.tables;
    assert_eq!(tables.len(), 2);
    assert!(tables.contains(&"table1".to_string()));
    assert!(tables.contains(&"table2".to_string()));
}

/// Test listing tables with non-root namespace ID should fail
#[tokio::test]
async fn test_list_tables_with_namespace_id() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    let mut request = ListTablesRequest::new();
    request.id = Some(vec!["namespace".to_string()]);

    let result = namespace.list_tables(request).await;
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_string()
        .contains("root namespace operations"));
}

/// Test describing a table
#[tokio::test]
async fn test_describe_table() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    // Create a table first
    let mut create_request = CreateTableRequest::new();
    create_request.id = Some(vec!["test_table".to_string()]);
    create_request.schema = Some(Box::new(create_test_schema()));
    namespace
        .create_table(create_request, bytes::Bytes::new())
        .await
        .unwrap();

    // Describe the table
    let mut request = DescribeTableRequest::new();
    request.id = Some(vec!["test_table".to_string()]);
    let response = namespace.describe_table(request).await.unwrap();
    
    assert!(response.location.is_some());
    assert!(response.location.unwrap().ends_with("test_table.lance"));
}

/// Test describing a non-existent table
#[tokio::test]
async fn test_describe_nonexistent_table() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    let mut request = DescribeTableRequest::new();
    request.id = Some(vec!["nonexistent".to_string()]);
    
    let result = namespace.describe_table(request).await;
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_string()
        .contains("Table does not exist"));
}

/// Test table_exists
#[tokio::test]
async fn test_table_exists() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    // Create a table
    let mut create_request = CreateTableRequest::new();
    create_request.id = Some(vec!["existing_table".to_string()]);
    create_request.schema = Some(Box::new(create_test_schema()));
    namespace
        .create_table(create_request, bytes::Bytes::new())
        .await
        .unwrap();

    // Check existing table
    let mut request = TableExistsRequest::new();
    request.id = Some(vec!["existing_table".to_string()]);
    let result = namespace.table_exists(request).await;
    assert!(result.is_ok());

    // Check non-existent table
    let mut request = TableExistsRequest::new();
    request.id = Some(vec!["nonexistent".to_string()]);
    let result = namespace.table_exists(request).await;
    assert!(result.is_err());
    assert!(result
        .unwrap_err()
        .to_string()
        .contains("Table does not exist"));
}

/// Test dropping a table
#[tokio::test]
async fn test_drop_table() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    // Create a table
    let mut create_request = CreateTableRequest::new();
    create_request.id = Some(vec!["table_to_drop".to_string()]);
    create_request.schema = Some(Box::new(create_test_schema()));
    namespace
        .create_table(create_request, bytes::Bytes::new())
        .await
        .unwrap();

    // Verify it exists
    let mut exists_request = TableExistsRequest::new();
    exists_request.id = Some(vec!["table_to_drop".to_string()]);
    assert!(namespace.table_exists(exists_request.clone()).await.is_ok());

    // Drop the table
    let mut drop_request = DropTableRequest::new();
    drop_request.id = Some(vec!["table_to_drop".to_string()]);
    let response = namespace.drop_table(drop_request).await.unwrap();
    assert!(response.location.is_some());

    // Verify it no longer exists
    assert!(namespace.table_exists(exists_request).await.is_err());
}

/// Test dropping a non-existent table
#[tokio::test]
async fn test_drop_nonexistent_table() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    let mut request = DropTableRequest::new();
    request.id = Some(vec!["nonexistent".to_string()]);
    
    // Should not fail when dropping non-existent table (idempotent)
    let result = namespace.drop_table(request).await;
    // The operation might succeed or fail depending on implementation
    // But it should not panic
    let _ = result;
}

/// Test namespace operations that are not supported
#[tokio::test]
async fn test_unsupported_namespace_operations() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    // Test create_namespace
    let request = lance_namespace_reqwest_client::models::CreateNamespaceRequest::new();
    let result = namespace.create_namespace(request).await;
    assert!(matches!(result, Err(NamespaceError::NotSupported(_))));

    // Test list_namespaces
    let request = lance_namespace_reqwest_client::models::ListNamespacesRequest::new();
    let result = namespace.list_namespaces(request).await;
    assert!(matches!(result, Err(NamespaceError::NotSupported(_))));

    // Test describe_namespace
    let request = lance_namespace_reqwest_client::models::DescribeNamespaceRequest::new();
    let result = namespace.describe_namespace(request).await;
    assert!(matches!(result, Err(NamespaceError::NotSupported(_))));

    // Test drop_namespace
    let request = lance_namespace_reqwest_client::models::DropNamespaceRequest::new();
    let result = namespace.drop_namespace(request).await;
    assert!(matches!(result, Err(NamespaceError::NotSupported(_))));

    // Test namespace_exists
    let request = lance_namespace_reqwest_client::models::NamespaceExistsRequest::new();
    let result = namespace.namespace_exists(request).await;
    assert!(matches!(result, Err(NamespaceError::NotSupported(_))));
}

/// Test configuration with custom root
#[tokio::test]
async fn test_config_custom_root() {
    let temp_dir = TempDir::new().unwrap();
    let custom_path = temp_dir.path().join("custom");
    std::fs::create_dir(&custom_path).unwrap();

    let mut properties = HashMap::new();
    properties.insert("root".to_string(), custom_path.to_string_lossy().to_string());

    let namespace = DirectoryNamespace::new(properties).unwrap();

    // Create a table and verify location
    let mut request = CreateTableRequest::new();
    request.id = Some(vec!["test_table".to_string()]);
    request.schema = Some(Box::new(create_test_schema()));

    let response = namespace
        .create_table(request, bytes::Bytes::new())
        .await
        .unwrap();

    assert!(response.location.unwrap().contains("custom"));
}

/// Test configuration with storage options
#[tokio::test]
async fn test_config_storage_options() {
    let temp_dir = TempDir::new().unwrap();
    let mut properties = HashMap::new();
    properties.insert(
        "root".to_string(),
        temp_dir.path().to_string_lossy().to_string(),
    );
    properties.insert("storage.option1".to_string(), "value1".to_string());
    properties.insert("storage.option2".to_string(), "value2".to_string());

    let namespace = DirectoryNamespace::new(properties).unwrap();

    // Create a table and check storage options are included
    let mut request = CreateTableRequest::new();
    request.id = Some(vec!["test_table".to_string()]);
    request.schema = Some(Box::new(create_test_schema()));

    let response = namespace
        .create_table(request, bytes::Bytes::new())
        .await
        .unwrap();

    let storage_options = response.storage_options.unwrap();
    assert_eq!(storage_options.get("option1"), Some(&"value1".to_string()));
    assert_eq!(storage_options.get("option2"), Some(&"value2".to_string()));
}

/// Test different Arrow data types in schema
#[tokio::test]
async fn test_various_arrow_types() {
    let (namespace, _temp_dir) = create_test_namespace().await;

    // Create schema with various types
    let mut fields = Vec::new();

    // Boolean field
    let bool_field = JsonArrowField {
        name: "bool_col".to_string(),
        r#type: Box::new(JsonArrowDataType::new("bool".to_string())),
        nullable: true,
        metadata: None,
    };
    fields.push(bool_field);

    // Int8 field
    let int8_field = JsonArrowField {
        name: "int8_col".to_string(),
        r#type: Box::new(JsonArrowDataType::new("int8".to_string())),
        nullable: true,
        metadata: None,
    };
    fields.push(int8_field);

    // Float64 field
    let float64_field = JsonArrowField {
        name: "float64_col".to_string(),
        r#type: Box::new(JsonArrowDataType::new("float64".to_string())),
        nullable: true,
        metadata: None,
    };
    fields.push(float64_field);

    // Binary field
    let binary_field = JsonArrowField {
        name: "binary_col".to_string(),
        r#type: Box::new(JsonArrowDataType::new("binary".to_string())),
        nullable: true,
        metadata: None,
    };
    fields.push(binary_field);

    let schema = JsonArrowSchema {
        fields,
        metadata: None,
    };

    let mut request = CreateTableRequest::new();
    request.id = Some(vec!["complex_table".to_string()]);
    request.schema = Some(Box::new(schema));

    let response = namespace
        .create_table(request, bytes::Bytes::new())
        .await
        .unwrap();

    assert!(response.location.is_some());
}

/// Test connect function with dir implementation
#[tokio::test]
async fn test_connect_dir() {
    let temp_dir = TempDir::new().unwrap();
    let mut properties = HashMap::new();
    properties.insert(
        "root".to_string(),
        temp_dir.path().to_string_lossy().to_string(),
    );

    let namespace = lance_namespace::connect("dir", properties)
        .await
        .unwrap();

    // Test basic operation through the trait object
    let request = ListTablesRequest::new();
    let response = namespace.list_tables(request).await.unwrap();
    assert_eq!(response.tables.len(), 0);
}